backend/globaleaks/jobs/x509_cert_check_sched.py

```diff
@@ -0,0 +1,106 @@
+from globaleaks.handlers.admin.https import db_acme_cert_issuance
+from globaleaks.handlers.admin.user import db_get_admin_users
+from globaleaks.jobs.base import GLJob
+from globaleaks.orm import transact_sync
+from globaleaks.security import encrypt_pgp_message
+from globaleaks.settings import GLSettings
+from globaleaks.utils import lets_enc
+from globaleaks.utils.templating import Templating
+from globaleaks.utils.utility import log
+
+
+def should_try_acme_renewal(failures):
+    # TODO check num failures isn't high
+    acme_autorenew = GLSettings.memory_copy.private.acme_autorenew
+    https_enabled = GLSettings.memory_copy.private.https_enabled
+    if https_enabled and acme_autorenew:
+        return True
+    return False
+
+
+class X509CertCheckSchedule(GLJob):
+    name = "X509 Cert Check"
+    interval = 3 * 24 * 3600
+
+    notify_expr_within = 15
+    acme_failures = 0
+    acme_try_renewal = 30
+
+    def operation(self):
+        if should_try_acme_renewal(self.acme_failures):
+             self.acme_cert_renewal_checks()
+        self.cert_expiration_checks()
+
+    @transact_sync
+    def acme_cert_renewal_checks(self, store):
+        cert = load_certificate(FILETYPE_PEM, GLSettings.memory_copy.private.https_cert)
+        expiration_date = lets_enc.convert_asn1_date(cert.get_notAfter())
+
+        t = timedelta(days=self.acme_try_renewal)
+        renewal_window = datetime.now() + t
+
+        if not expiration_date < renewal_window:
+            # We will not apply for the renewal of the certificate
+            return
+
+        try:
+            db_acme_cert_issuance(store, None)
+        except Exception as e:
+            self.acme_failures =+ 1
+            log.err('ACME certificate renewal failed with: %s' % e)
+            raise
+        try:
+            yield GLSettings.state.process_supervisor.shutdown()
+            yield GLSettings.state.process_supervisor.maybe_launch_https_workers()
+        except Exception as e:
+            self.acme_failures =+ 1
+            log.err('Restart of HTTPS workers failed with: %s' % e)
+            raise
+
+    @transact_sync
+    def cert_expiration_checks(self, store):
+        if not GLSettings.memory_copy.private.https_enabled:
+            return
+
+        cert = load_certificate(FILETYPE_PEM, GLSettings.memory_copy.private.https_cert)
+        expiration_date = lets_enc.convert_asn1_date(cert.get_notAfter())
+
+        t = timedelta(days=self.notify_expr_within)
+        expiration_window = datetime.now() + t
+        if not expiration_date < expiration_window:
+            log.debug('The HTTPS certificate is not going to expire within target window.')
+            return
+
+        # The certificate is going to expire. Mail the administrator
+        self.certificate_mail_creation(store, expiration_date)
+
+    def certificate_mail_creation(self, store, expiration_date):
+        for user_desc in db_get_admin_users(store):
```

Comment:
on the stile for the new code to be kept in mind just for the future:

maybe in general it will be possible to simplify this for loops on users so to remove the need for nested code.

this in general as direction for having more small functions and improve the code quality